File: python-workers/models/rules.py
from sqlalchemy import UUID, Column, Integer, String, Text, Boolean, DateTime, Time, func
from config.database import Base

class Rules(Base):
    __tablename__ = 'rules'

    id = Column(UUID, primary_key=True)
    name = Column(String, nullable=False)
    description = Column(Text)
    database_type = Column(String)
    sql = Column(Text)
    priority = Column(String)
    execution_interval_ms = Column(Integer)
    max_error_count = Column(Integer)
    timeout_ms = Column(Integer)
    start_time = Column(Time)
    end_time = Column(Time)
    is_active = Column(Boolean, default=True)
    silence_mode = Column(Boolean, default=False)
    postpone_date = Column(DateTime)
    user_creator_id = Column(UUID)
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now())

    roles = None

    @staticmethod
    def from_array(rules_array):
        return [Rules(**rule) if isinstance(rule, dict) else rule for rule in rules_array]

File: qqtech9-backend/src/models/rules.js
import { BusinessLogicError } from "../utils/errors.js";
import { validateSQLQueryAST, validateSQLQueryRegex } from '../utils/sql-validation.js';

export class Rules {
    constructor(rule) {
        this.id = rule.id;
        this.name = rule.name;
        this.description = rule.description;
        this.databaseType = rule.database_type ?? rule.databaseType;
        this.sql = rule.sql;
        this.priority = rule.priority;
        this.roles = rule.roles;
        this.executionIntervalMs = rule.execution_interval_ms ?? rule.executionIntervalMs;
        this.maxErrorCount = rule.max_error_count ?? rule.maxErrorCount;
        this.timeoutMs = rule.timeout_ms ?? rule.timeoutMs;
        this.startTime = rule.start_time ?? rule.startTime;
        this.endTime = rule.end_time ?? rule.endTime;
        this.isActive = rule.is_active ?? rule.isActive;
        this.silenceMode = rule.silence_mode ?? rule.silenceMode;
        this.postponeDate = rule.postpone_date ?? rule.postponeDate;
        this.userCreatorId = rule.user_creator_id ?? rule.userCreatorId;
        this.createdAt = rule.created_at ?? rule.createdAt;
        this.updatedAt = rule.updated_at ?? rule.updatedAt;
    }

    static fromArray(rulesArray) {
        return rulesArray.map((rule) => new Rules(rule));
    }

    reexecute() {
        if(!this.isActive){
            this.isActive = true;
        }

        return this;
    }

    validateBusinessLogic() {
        if (!validateSQLQueryRegex(this.sql)) {
            throw new BusinessLogicError('SQL contains forbidden commands');
        }
        if (validateSQLQueryAST(this.sql).valid === false) {
            throw new BusinessLogicError(`${validateSQLQueryAST(this.sql).error}`);
        }
        if (this.postponeDate && new Date(this.postponeDate) < new Date()) {
            throw new BusinessLogicError('Postpone date must be in the future');
        }
        if (this.executionIntervalMs <= 0) {
            throw new BusinessLogicError('Execution interval must be positive');
        }
        if (this.maxErrorCount <= 0) {
            throw new BusinessLogicError('Max error must be positive');
        }
        if (this.timeoutMs <= 0) {
            throw new BusinessLogicError('Timeout must be positive');
        }
        
        return this;
    }
}